src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs

```diff
@@ -491,12 +491,45 @@ async Task<PurgeHistoryResult> IDurableOrchestrationClient.PurgeInstanceHistoryA
             return this.DurabilityProvider.GetOrchestrationStateWithPagination(condition, cancellationToken);
         }
 
+        private static EntityQueryResult ConvertToEntityQueryResult(IEnumerable<DurableEntityStatus> entities, string continuationToken)
+        {
+            var entityQueryResult = new EntityQueryResult();
+            entityQueryResult.Entities = entities;
+            entityQueryResult.ContinuationToken = continuationToken;
+
+            return entityQueryResult;
+        }
+
         /// <inheritdoc />
         async Task<EntityQueryResult> IDurableEntityClient.ListEntitiesAsync(EntityQuery query, CancellationToken cancellationToken)
         {
+            if (!query.IncludeDeleted)
+            {
+                query.FetchState = true;
+            }
+
             var condition = new OrchestrationStatusQueryCondition(query);
             var result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
             var entityResult = new EntityQueryResult(result);
+
+            if (!query.IncludeDeleted)
+            {
+                var statefulEntities = entityResult.Entities.Where(e => e.State != null);
+                var statefulEntityResult = ConvertToEntityQueryResult(statefulEntities, entityResult.ContinuationToken);
+
+                while (statefulEntities.ToList().Count < query.PageSize && !string.IsNullOrEmpty(entityResult.ContinuationToken))
+                {
+                    query.ContinuationToken = entityResult.ContinuationToken;
+                    condition = new OrchestrationStatusQueryCondition(query);
+                    result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
+                    entityResult = new EntityQueryResult(result);
+                    statefulEntities = entityResult.Entities.Where(e => e.State != null);
```

Comment:
Since you're overwriting `statefulEntities` here, aren't you also overwriting any of the previously fetched results? For implementing this task, I would have expected you to have a `List<DurableEntityStatus> statefulEntities` outside the loop and then call `statefulEntities.AddRange(...)` inside of every loop iteration until the size of the list reaches the `query.PageSize`.